{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "622ecccb",
   "metadata": {},
   "source": [
    "# csaQTL testing in datasets with suggestive trans-eGenes removed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d31ee410",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pickle, cna\n",
    "import scipy.stats as st"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a4d89089",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>celltype</th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>A1</th>\n",
       "      <th>TEST</th>\n",
       "      <th>OBS_CT</th>\n",
       "      <th>P</th>\n",
       "      <th>...</th>\n",
       "      <th>CHR</th>\n",
       "      <th>MAF</th>\n",
       "      <th>Cell Type</th>\n",
       "      <th>SNP</th>\n",
       "      <th>Effect Allele</th>\n",
       "      <th>Other Allele</th>\n",
       "      <th>rsID</th>\n",
       "      <th>Cyto. Band</th>\n",
       "      <th>Phenotype Annotation</th>\n",
       "      <th>Novel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NK</td>\n",
       "      <td>2</td>\n",
       "      <td>111851212</td>\n",
       "      <td>2:111851212:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>ADD</td>\n",
       "      <td>935</td>\n",
       "      <td>1.764828e-09</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0.21504</td>\n",
       "      <td>NK</td>\n",
       "      <td>2:111851212:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs13025330</td>\n",
       "      <td>2q13</td>\n",
       "      <td>Decrease in CD56bright \\% NK cells</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NK</td>\n",
       "      <td>11</td>\n",
       "      <td>128070535</td>\n",
       "      <td>11:128070535:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>ADD</td>\n",
       "      <td>935</td>\n",
       "      <td>2.480238e-13</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>0.27872</td>\n",
       "      <td>NK</td>\n",
       "      <td>11:128070535:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs519062</td>\n",
       "      <td>11q24.3</td>\n",
       "      <td>Increase in NK cells activated by TNF-$\\alpha$...</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NK</td>\n",
       "      <td>12</td>\n",
       "      <td>10583611</td>\n",
       "      <td>12:10583611:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ADD</td>\n",
       "      <td>935</td>\n",
       "      <td>1.956080e-11</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>0.17811</td>\n",
       "      <td>NK</td>\n",
       "      <td>12:10583611:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs3003</td>\n",
       "      <td>12p13.2</td>\n",
       "      <td>Increase in NK cells activated by TNF-$\\alpha$...</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NK</td>\n",
       "      <td>19</td>\n",
       "      <td>16441973</td>\n",
       "      <td>19:16441973:G:A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>935</td>\n",
       "      <td>1.958433e-13</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>0.32867</td>\n",
       "      <td>NK</td>\n",
       "      <td>19:16441973:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>rs56133626</td>\n",
       "      <td>19p13.11</td>\n",
       "      <td>Increase in NK cells activated by TNF-$\\alpha$...</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  celltype  #CHROM        POS                ID REF ALT A1 TEST  OBS_CT  \\\n",
       "0       NK       2  111851212   2:111851212:C:T   C   T  T  ADD     935   \n",
       "1       NK      11  128070535  11:128070535:A:G   A   G  G  ADD     935   \n",
       "2       NK      12   10583611   12:10583611:C:T   C   T  C  ADD     935   \n",
       "3       NK      19   16441973   19:16441973:G:A   G   A  A  ADD     935   \n",
       "\n",
       "              P  ...  CHR      MAF  Cell Type               SNP  \\\n",
       "0  1.764828e-09  ...    2  0.21504         NK   2:111851212:C:T   \n",
       "1  2.480238e-13  ...   11  0.27872         NK  11:128070535:A:G   \n",
       "2  1.956080e-11  ...   12  0.17811         NK   12:10583611:C:T   \n",
       "3  1.958433e-13  ...   19  0.32867         NK   19:16441973:G:A   \n",
       "\n",
       "   Effect Allele  Other Allele        rsID  Cyto. Band  \\\n",
       "0              T             C  rs13025330        2q13   \n",
       "1              G             A    rs519062     11q24.3   \n",
       "2              T             C      rs3003     12p13.2   \n",
       "3              A             G  rs56133626    19p13.11   \n",
       "\n",
       "                                Phenotype Annotation  Novel  \n",
       "0                 Decrease in CD56bright \\% NK cells     No  \n",
       "1  Increase in NK cells activated by TNF-$\\alpha$...    Yes  \n",
       "2  Increase in NK cells activated by TNF-$\\alpha$...    Yes  \n",
       "3  Increase in NK cells activated by TNF-$\\alpha$...    Yes  \n",
       "\n",
       "[4 rows x 46 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci = pd.read_csv(\"/data/srlab/lrumker/MCSC_Project/cna-qtl/results/loci_table.tsv\", sep = \"\\t\")\n",
    "loci = loci.iloc[1:,:].reset_index(drop=True) # no trans eQTL candidates for the myeloid csaQTL\n",
    "loci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bb495e0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "eQTL_dep_table = pd.DataFrame({})\n",
    "for i_locus in np.arange(loci.shape[0]):\n",
    "    celltype = loci['celltype'][i_locus]\n",
    "    lead_snp = loci['ID'][i_locus]\n",
    "    gwas_p = '{:.2e}'.format(loci.iloc[i_locus, :].P)\n",
    "\n",
    "    masked_locus = pd.read_table('/data/srlab/lrumker/MCSC_Project/cna-qtl/mask_trans_eGenes/'+lead_snp+\\\n",
    "                                 \"/GeNA_sumstats.txt\")\n",
    "    masked_p = '{:.2e}'.format(masked_locus.loc[masked_locus.ID==lead_snp].P.values[0])\n",
    "\n",
    "    # load data\n",
    "    res = pickle.load(open('/data/srlab/lrumker/MCSC_Project/cna-qtl/results/gwas_'+\\\n",
    "                                   celltype+\"/cna_res_\"+lead_snp+'.p','rb'))\n",
    "    res_mask = pickle.load(open('/data/srlab/lrumker/MCSC_Project/cna-qtl/mask_trans_eGenes/'+\\\n",
    "                                celltype+'_'+lead_snp+'_mask_trans_eGenes.p','rb'))\n",
    "\n",
    "    # trans eGenes removed\n",
    "    dropped = [i for i in np.arange(res.vargene_cors.shape[0]) if \n",
    "                          res.vargene_cors.gene[i] not in res_mask.vargene_cors.gene.values]\n",
    "    dropped_genes = res.vargene_cors.gene.values[dropped]\n",
    "\n",
    "    # expression change in these genes R\n",
    "    dropped_cors = res.vargene_cors.iloc[dropped,:].cor.values\n",
    "\n",
    "    # percentile of R^2 among all genes\n",
    "    rsq_all_vargenes = res.vargene_cors.cor.values**2\n",
    "    pctile_rsq = [np.sum(rsq_all_vargenes<(dropped_cors[i]**2))/res.vargene_cors.shape[0]\\\n",
    "                    for i in np.arange(len(dropped_genes))]\n",
    "    pctile_rsq = np.around(pctile_rsq,2)\n",
    "\n",
    "    # sort genes\n",
    "    sel_order = np.argsort(-pctile_rsq)\n",
    "    droppped_genes_str = \", \".join(dropped_genes[sel_order])\n",
    "    droppped_cors_str = \", \".join(np.around(dropped_cors[sel_order],3).astype(str))\n",
    "    pctile_rsq_str = \", \".join(((pctile_rsq[sel_order]*100).astype(int)).astype(str))\n",
    "\n",
    "    res.vargene_cors.set_index('gene', inplace = True, drop = True)\n",
    "    res_mask.vargene_cors.set_index('gene', inplace = True, drop = True)\n",
    "\n",
    "    # correlation in pre/post vargene correlations to ncorrs (retained genes)\n",
    "    Rsq_nbhd = np.around(np.corrcoef(res.vargene_cors.loc[res_mask.vargene_cors.index,:].cor, \n",
    "                          res_mask.vargene_cors.cor )[0,1]**2,2)\n",
    "\n",
    "    # correlation in pre/post phenotypes, per individual, Rsq\n",
    "    d = cna.read(\"/data/srlab/lrumker/datasets/onek1k/pheno/\"+celltype+\".h5ad\")\n",
    "    d_mc = cna.read(\"/data/srlab/lrumker/MCSC_Project/cna-qtl/mask_trans_eGenes/masked_\"+\\\n",
    "                    celltype+\"_\"+lead_snp+\".h5ad\")\n",
    "\n",
    "    nampcs = d.uns['NAM_sampleXpc'].iloc[:,:res.k]\n",
    "    pheno = np.dot(res.beta.reshape(1,-1), nampcs.T).reshape(-1,)\n",
    "    nampcs = d_mc.uns['NAM_sampleXpc'].iloc[:,:res_mask.k]\n",
    "    pheno_mask = np.dot(nampcs, res_mask.beta)\n",
    "    Rsq_sample = np.around(np.corrcoef(pheno, pheno_mask)[0,1]**2,2)\n",
    "\n",
    "    eQTL_dep_table = pd.concat((eQTL_dep_table,\n",
    "                                pd.DataFrame({\"Celltype\": [celltype], \"SNP\": [lead_snp], \"Cis Vargenes\": [droppped_genes_str],\n",
    "                 \"Expr Cor (R)\": [droppped_cors_str], \"%ile Expr Cor (Rsq)\": [pctile_rsq_str],\n",
    "                 \"Rsq Pheno, Nbhds\": [Rsq_nbhd], \"Rsq Pheno, Samples\": Rsq_sample,\n",
    "                                             \"GWAS P\": gwas_p, \"P, Masked\": masked_p})))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "26864a84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# No candidate trans-eQTLs detected for the myeloid csaQTL\n",
    "eQTL_pvals=[]\n",
    "for i in np.arange(eQTL_dep_table.shape[0]):\n",
    "    celltype=eQTL_dep_table.Celltype.values[i]\n",
    "    lead_snp=eQTL_dep_table.SNP.values[i]\n",
    "    genes=eQTL_dep_table['Cis Vargenes'].values[i].split(\", \") # This column label is later replaced with Trans-eGenes\n",
    "    \n",
    "    # Note that not all vargenes passed PEER QC for pseudobulk eQTL testing\n",
    "    res = pd.read_csv(\"/data/srlab/lrumker/MCSC_Project/cna-qtl/eqtls/results/trans_eQTL_candidates/pseudobulk/\"+\n",
    "                  celltype+\"_\"+lead_snp+\"_csaQTL_test_eQTLs_pseudobulk_eQTLs.csv\", index_col = 0)\n",
    "    res = res.loc[res.variant.values==lead_snp,:].reset_index(drop=True)\n",
    "    res.set_index('gene', inplace = True)\n",
    "    new_pvals = res.loc[genes,'p.val'].values\n",
    "    new_pvals = ['{:.2e}'.format(new_pvals[j]) for j in np.arange(len(new_pvals))]\n",
    "    new_pvals = \", \".join(new_pvals)\n",
    "    eQTL_pvals.append(new_pvals)\n",
    "    \n",
    "eQTL_dep_table['eQTL P, Pseudobulk'] = eQTL_pvals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a975707a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Celltype</th>\n",
       "      <th>SNP</th>\n",
       "      <th>Cis Vargenes</th>\n",
       "      <th>Expr Cor (R)</th>\n",
       "      <th>%ile Expr Cor (Rsq)</th>\n",
       "      <th>Rsq Pheno, Nbhds</th>\n",
       "      <th>Rsq Pheno, Samples</th>\n",
       "      <th>GWAS P</th>\n",
       "      <th>P, Masked</th>\n",
       "      <th>eQTL P, Pseudobulk</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NK</td>\n",
       "      <td>2:111851212:C:T</td>\n",
       "      <td>GZMK</td>\n",
       "      <td>-0.525</td>\n",
       "      <td>100</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.76e-09</td>\n",
       "      <td>6.70e-09</td>\n",
       "      <td>3.14e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NK</td>\n",
       "      <td>11:128070535:A:G</td>\n",
       "      <td>PRSS23, TMIGD2, CNFN</td>\n",
       "      <td>-0.204, -0.114, -0.106</td>\n",
       "      <td>100, 98, 97</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.99</td>\n",
       "      <td>2.48e-13</td>\n",
       "      <td>2.63e-12</td>\n",
       "      <td>1.80e-08, 5.17e-06, 3.81e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NK</td>\n",
       "      <td>12:10583611:C:T</td>\n",
       "      <td>CD52</td>\n",
       "      <td>-0.204</td>\n",
       "      <td>100</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.95</td>\n",
       "      <td>1.96e-11</td>\n",
       "      <td>4.67e-10</td>\n",
       "      <td>2.42e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NK</td>\n",
       "      <td>19:16441973:G:A</td>\n",
       "      <td>PRSS23, SH3BGRL3, CD52, ZNF683, CNFN, ANXA1, C...</td>\n",
       "      <td>-0.28, -0.239, -0.223, -0.244, -0.118, -0.107,...</td>\n",
       "      <td>100, 99, 99, 99, 97, 96, 94</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.93</td>\n",
       "      <td>1.96e-13</td>\n",
       "      <td>3.53e-11</td>\n",
       "      <td>3.20e-07, 6.17e-07, 3.63e-05, 3.40e-09, 5.82e-...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Celltype               SNP  \\\n",
       "0       NK   2:111851212:C:T   \n",
       "0       NK  11:128070535:A:G   \n",
       "0       NK   12:10583611:C:T   \n",
       "0       NK   19:16441973:G:A   \n",
       "\n",
       "                                        Cis Vargenes  \\\n",
       "0                                               GZMK   \n",
       "0                               PRSS23, TMIGD2, CNFN   \n",
       "0                                               CD52   \n",
       "0  PRSS23, SH3BGRL3, CD52, ZNF683, CNFN, ANXA1, C...   \n",
       "\n",
       "                                        Expr Cor (R)  \\\n",
       "0                                             -0.525   \n",
       "0                             -0.204, -0.114, -0.106   \n",
       "0                                             -0.204   \n",
       "0  -0.28, -0.239, -0.223, -0.244, -0.118, -0.107,...   \n",
       "\n",
       "           %ile Expr Cor (Rsq)  Rsq Pheno, Nbhds  Rsq Pheno, Samples  \\\n",
       "0                          100              1.00                1.00   \n",
       "0                  100, 98, 97              1.00                0.99   \n",
       "0                          100              0.99                0.95   \n",
       "0  100, 99, 99, 99, 97, 96, 94              0.99                0.93   \n",
       "\n",
       "     GWAS P P, Masked                                 eQTL P, Pseudobulk  \n",
       "0  1.76e-09  6.70e-09                                           3.14e-06  \n",
       "0  2.48e-13  2.63e-12                       1.80e-08, 5.17e-06, 3.81e-05  \n",
       "0  1.96e-11  4.67e-10                                           2.42e-05  \n",
       "0  1.96e-13  3.53e-11  3.20e-07, 6.17e-07, 3.63e-05, 3.40e-09, 5.82e-...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_folder=\"/data/srlab/lrumker/MCSC_Project/cna-qtl/results/\"\n",
    "eQTL_dep_table.to_csv(res_folder+\"trans_eQTL_dep_table.tsv\", sep = \"\\t\", index = False)\n",
    "eQTL_dep_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "801bba00",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Format results for exported table\n",
    "res_folder=\"/data/srlab/lrumker/MCSC_Project/cna-qtl/results/\"\n",
    "eQTL_dep_table = pd.read_csv(res_folder+\"trans_eQTL_dep_table.tsv\", sep = \"\\t\")\n",
    "eQTL_dep_table['GWAS P'] = [str(eQTL_dep_table['GWAS P'].values[i]) for i in np.arange(eQTL_dep_table.shape[0])]\n",
    "eQTL_dep_table['P, Masked'] = [str(eQTL_dep_table['P, Masked'].values[i]) for i in np.arange(eQTL_dep_table.shape[0])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fc8a0a6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "eQTL_dep_table['Cis Vargenes'] = [\", \".join([r\"\\textit{\"+eQTL_dep_table['Cis Vargenes'].values[j].split(\",\")[i]+r\"}\" \\\n",
    " for i in np.arange(len(eQTL_dep_table['Cis Vargenes'].values[j].split(\",\")))]) \\\n",
    "                                 for j in np.arange(eQTL_dep_table.shape[0])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "07074bbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# \\shortstack{a \\\\ bb \\\\ c} for vertical stacking within one table cell\n",
    "eQTL_dep_table['%ile Expr Cor (Rsq)'] = [r\"\\shortstack{\"+r' \\\\ '.join(eQTL_dep_table['%ile Expr Cor (Rsq)'].values[i].split(\", \"))+\"}\" \\\n",
    "                                  if (type(eQTL_dep_table['%ile Expr Cor (Rsq)'].values[i]) is not float \\\n",
    "                                      and len(eQTL_dep_table['%ile Expr Cor (Rsq)'].values[i].split(\", \")) != 1 ) \\\n",
    "                                        else eQTL_dep_table['%ile Expr Cor (Rsq)'].values[i] \\\n",
    "                                         for i in np.arange(eQTL_dep_table.shape[0]) ]\n",
    "eQTL_dep_table['Expr Cor (R)'] = [r\"\\shortstack{\"+r' \\\\ '.join(eQTL_dep_table['Expr Cor (R)'].values[i].split(\", \"))+\"}\" \\\n",
    "                                  if (type(eQTL_dep_table['Expr Cor (R)'].values[i]) is not float \\\n",
    "                                      and len(eQTL_dep_table['Expr Cor (R)'].values[i].split(\", \")) != 1 ) \\\n",
    "                                        else eQTL_dep_table['Expr Cor (R)'].values[i] \\\n",
    "                                  for i in np.arange(eQTL_dep_table.shape[0])]\n",
    "eQTL_dep_table['Cis Vargenes'] = [r\"\\shortstack{\"+r' \\\\ '.join(eQTL_dep_table['Cis Vargenes'].values[i].split(\", \"))+\"}\" \\\n",
    "                                  if (type(eQTL_dep_table['Cis Vargenes'].values[i]) is not float \\\n",
    "                                      and len(eQTL_dep_table['Cis Vargenes'].values[i].split(\", \")) != 1 ) \\\n",
    "                                        else eQTL_dep_table['Cis Vargenes'].values[i] \\\n",
    "                                  for i in np.arange(eQTL_dep_table.shape[0])]\n",
    "\n",
    "eQTL_dep_table['eQTL P, Pseudobulk'] = [r\"\\shortstack{\"+r' \\\\ '.join(eQTL_dep_table['eQTL P, Pseudobulk'].values[i].split(\", \"))+\"}\" \\\n",
    "                                  if (type(eQTL_dep_table['eQTL P, Pseudobulk'].values[i]) is not float \\\n",
    "                                      and len(eQTL_dep_table['eQTL P, Pseudobulk'].values[i].split(\", \")) != 1 ) \\\n",
    "                                        else eQTL_dep_table['eQTL P, Pseudobulk'].values[i] \\\n",
    "                                  for i in np.arange(eQTL_dep_table.shape[0])]\n",
    "savenames = eQTL_dep_table.columns\n",
    "savenames = [savenames[i] if i !=0 else 'Cell Type' for i in np.arange(len(savenames))]\n",
    "savenames = [savenames[i] if i !=1 else 'CHR:POS' for i in np.arange(len(savenames))]\n",
    "savenames = [savenames[i] if i !=2 else 'Trans-eGenes' for i in np.arange(len(savenames))]\n",
    "savenames = [savenames[i] if i !=3 else 'Expr Cor (r)' for i in np.arange(len(savenames))]\n",
    "savenames = [savenames[i] if i !=4 else r'\\%tile Expr Cor (r$^{2}$)' for i in np.arange(len(savenames))]\n",
    "savenames = [savenames[i] if i !=5 else 'r$^{2}$ Pheno, Nbhds' for i in np.arange(len(savenames))]\n",
    "savenames = [savenames[i] if i !=6 else 'r$^{2}$ Pheno, Samples' for i in np.arange(len(savenames))]\n",
    "eQTL_dep_table.columns = savenames\n",
    "\n",
    "eQTL_dep_table['CHR:POS'] = [\":\".join(eQTL_dep_table['CHR:POS'].values[i].split(\":\")[:2]) \\\n",
    "                             for i in np.arange(eQTL_dep_table.shape[0])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c3ed0490",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cell Type</th>\n",
       "      <th>CHR:POS</th>\n",
       "      <th>Trans-eGenes</th>\n",
       "      <th>eQTL P, Pseudobulk</th>\n",
       "      <th>Expr Cor (r)</th>\n",
       "      <th>\\%tile Expr Cor (r$^{2}$)</th>\n",
       "      <th>r$^{2}$ Pheno, Nbhds</th>\n",
       "      <th>r$^{2}$ Pheno, Samples</th>\n",
       "      <th>GWAS P</th>\n",
       "      <th>P, Masked</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NK</td>\n",
       "      <td>2:111851212</td>\n",
       "      <td>\\textit{GZMK}</td>\n",
       "      <td>3.14e-06</td>\n",
       "      <td>-0.525</td>\n",
       "      <td>100</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.76e-09</td>\n",
       "      <td>6.7e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NK</td>\n",
       "      <td>11:128070535</td>\n",
       "      <td>\\shortstack{\\textit{PRSS23} \\\\ \\textit{ TMIGD2...</td>\n",
       "      <td>\\shortstack{1.80e-08 \\\\ 5.17e-06 \\\\ 3.81e-05}</td>\n",
       "      <td>\\shortstack{-0.204 \\\\ -0.114 \\\\ -0.106}</td>\n",
       "      <td>\\shortstack{100 \\\\ 98 \\\\ 97}</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.99</td>\n",
       "      <td>2.48e-13</td>\n",
       "      <td>2.63e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NK</td>\n",
       "      <td>12:10583611</td>\n",
       "      <td>\\textit{CD52}</td>\n",
       "      <td>2.42e-05</td>\n",
       "      <td>-0.204</td>\n",
       "      <td>100</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.95</td>\n",
       "      <td>1.96e-11</td>\n",
       "      <td>4.67e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NK</td>\n",
       "      <td>19:16441973</td>\n",
       "      <td>\\shortstack{\\textit{PRSS23} \\\\ \\textit{ SH3BGR...</td>\n",
       "      <td>\\shortstack{3.20e-07 \\\\ 6.17e-07 \\\\ 3.63e-05 \\...</td>\n",
       "      <td>\\shortstack{-0.28 \\\\ -0.239 \\\\ -0.223 \\\\ -0.24...</td>\n",
       "      <td>\\shortstack{100 \\\\ 99 \\\\ 99 \\\\ 99 \\\\ 97 \\\\ 96 ...</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.93</td>\n",
       "      <td>1.96e-13</td>\n",
       "      <td>3.53e-11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Cell Type       CHR:POS                                       Trans-eGenes  \\\n",
       "0        NK   2:111851212                                      \\textit{GZMK}   \n",
       "1        NK  11:128070535  \\shortstack{\\textit{PRSS23} \\\\ \\textit{ TMIGD2...   \n",
       "2        NK   12:10583611                                      \\textit{CD52}   \n",
       "3        NK   19:16441973  \\shortstack{\\textit{PRSS23} \\\\ \\textit{ SH3BGR...   \n",
       "\n",
       "                                  eQTL P, Pseudobulk  \\\n",
       "0                                           3.14e-06   \n",
       "1      \\shortstack{1.80e-08 \\\\ 5.17e-06 \\\\ 3.81e-05}   \n",
       "2                                           2.42e-05   \n",
       "3  \\shortstack{3.20e-07 \\\\ 6.17e-07 \\\\ 3.63e-05 \\...   \n",
       "\n",
       "                                        Expr Cor (r)  \\\n",
       "0                                             -0.525   \n",
       "1            \\shortstack{-0.204 \\\\ -0.114 \\\\ -0.106}   \n",
       "2                                             -0.204   \n",
       "3  \\shortstack{-0.28 \\\\ -0.239 \\\\ -0.223 \\\\ -0.24...   \n",
       "\n",
       "                           \\%tile Expr Cor (r$^{2}$)  r$^{2}$ Pheno, Nbhds  \\\n",
       "0                                                100                  1.00   \n",
       "1                       \\shortstack{100 \\\\ 98 \\\\ 97}                  1.00   \n",
       "2                                                100                  0.99   \n",
       "3  \\shortstack{100 \\\\ 99 \\\\ 99 \\\\ 99 \\\\ 97 \\\\ 96 ...                  0.99   \n",
       "\n",
       "   r$^{2}$ Pheno, Samples    GWAS P P, Masked  \n",
       "0                    1.00  1.76e-09   6.7e-09  \n",
       "1                    0.99  2.48e-13  2.63e-12  \n",
       "2                    0.95  1.96e-11  4.67e-10  \n",
       "3                    0.93  1.96e-13  3.53e-11  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eQTL_dep_table = eQTL_dep_table.loc[:,['Cell Type', 'CHR:POS', 'Trans-eGenes', 'eQTL P, Pseudobulk',\n",
    "                     'Expr Cor (r)', r'\\%tile Expr Cor (r$^{2}$)', 'r$^{2}$ Pheno, Nbhds', \n",
    "                     'r$^{2}$ Pheno, Samples', 'GWAS P', 'P, Masked']]\n",
    "eQTL_dep_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "745c9ca0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cell Type</th>\n",
       "      <th>CHR:POS</th>\n",
       "      <th>Trans-eGenes</th>\n",
       "      <th>eQTL P, Pseudobulk</th>\n",
       "      <th>Expr Cor (r)</th>\n",
       "      <th>\\%tile Expr Cor (r$^{2}$)</th>\n",
       "      <th>r$^{2}$ Pheno, Nbhds</th>\n",
       "      <th>r$^{2}$ Pheno, Samples</th>\n",
       "      <th>GWAS P</th>\n",
       "      <th>P, Masked</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NK</td>\n",
       "      <td>2:111851212</td>\n",
       "      <td>\\textit{GZMK}</td>\n",
       "      <td>3.14e-06</td>\n",
       "      <td>-0.525</td>\n",
       "      <td>100</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.76e-09</td>\n",
       "      <td>6.7e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NK</td>\n",
       "      <td>11:128070535</td>\n",
       "      <td>\\shortstack{\\textit{PRSS23} \\\\ \\textit{ TMIGD2} \\\\ \\textit{ CNFN}}</td>\n",
       "      <td>\\shortstack{1.80e-08 \\\\ 5.17e-06 \\\\ 3.81e-05}</td>\n",
       "      <td>\\shortstack{-0.204 \\\\ -0.114 \\\\ -0.106}</td>\n",
       "      <td>\\shortstack{100 \\\\ 98 \\\\ 97}</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.99</td>\n",
       "      <td>2.48e-13</td>\n",
       "      <td>2.63e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NK</td>\n",
       "      <td>12:10583611</td>\n",
       "      <td>\\textit{CD52}</td>\n",
       "      <td>2.42e-05</td>\n",
       "      <td>-0.204</td>\n",
       "      <td>100</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.95</td>\n",
       "      <td>1.96e-11</td>\n",
       "      <td>4.67e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NK</td>\n",
       "      <td>19:16441973</td>\n",
       "      <td>\\shortstack{\\textit{PRSS23} \\\\ \\textit{ SH3BGRL3} \\\\ \\textit{ CD52} \\\\ \\textit{ ZNF683} \\\\ \\textit{ CNFN} \\\\ \\textit{ ANXA1} \\\\ \\textit{ COL13A1}}</td>\n",
       "      <td>\\shortstack{3.20e-07 \\\\ 6.17e-07 \\\\ 3.63e-05 \\\\ 3.40e-09 \\\\ 5.82e-07 \\\\ 2.00e-05 \\\\ 1.09e-05}</td>\n",
       "      <td>\\shortstack{-0.28 \\\\ -0.239 \\\\ -0.223 \\\\ -0.244 \\\\ -0.118 \\\\ -0.107 \\\\ -0.082}</td>\n",
       "      <td>\\shortstack{100 \\\\ 99 \\\\ 99 \\\\ 99 \\\\ 97 \\\\ 96 \\\\ 94}</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.93</td>\n",
       "      <td>1.96e-13</td>\n",
       "      <td>3.53e-11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Cell Type       CHR:POS  \\\n",
       "0        NK   2:111851212   \n",
       "1        NK  11:128070535   \n",
       "2        NK   12:10583611   \n",
       "3        NK   19:16441973   \n",
       "\n",
       "                                                                                                                                         Trans-eGenes  \\\n",
       "0                                                                                                                                       \\textit{GZMK}   \n",
       "1                                                                                  \\shortstack{\\textit{PRSS23} \\\\ \\textit{ TMIGD2} \\\\ \\textit{ CNFN}}   \n",
       "2                                                                                                                                       \\textit{CD52}   \n",
       "3  \\shortstack{\\textit{PRSS23} \\\\ \\textit{ SH3BGRL3} \\\\ \\textit{ CD52} \\\\ \\textit{ ZNF683} \\\\ \\textit{ CNFN} \\\\ \\textit{ ANXA1} \\\\ \\textit{ COL13A1}}   \n",
       "\n",
       "                                                                              eQTL P, Pseudobulk  \\\n",
       "0                                                                                       3.14e-06   \n",
       "1                                                  \\shortstack{1.80e-08 \\\\ 5.17e-06 \\\\ 3.81e-05}   \n",
       "2                                                                                       2.42e-05   \n",
       "3  \\shortstack{3.20e-07 \\\\ 6.17e-07 \\\\ 3.63e-05 \\\\ 3.40e-09 \\\\ 5.82e-07 \\\\ 2.00e-05 \\\\ 1.09e-05}   \n",
       "\n",
       "                                                                     Expr Cor (r)  \\\n",
       "0                                                                          -0.525   \n",
       "1                                         \\shortstack{-0.204 \\\\ -0.114 \\\\ -0.106}   \n",
       "2                                                                          -0.204   \n",
       "3  \\shortstack{-0.28 \\\\ -0.239 \\\\ -0.223 \\\\ -0.244 \\\\ -0.118 \\\\ -0.107 \\\\ -0.082}   \n",
       "\n",
       "                              \\%tile Expr Cor (r$^{2}$)  r$^{2}$ Pheno, Nbhds  \\\n",
       "0                                                   100                  1.00   \n",
       "1                          \\shortstack{100 \\\\ 98 \\\\ 97}                  1.00   \n",
       "2                                                   100                  0.99   \n",
       "3  \\shortstack{100 \\\\ 99 \\\\ 99 \\\\ 99 \\\\ 97 \\\\ 96 \\\\ 94}                  0.99   \n",
       "\n",
       "   r$^{2}$ Pheno, Samples    GWAS P P, Masked  \n",
       "0                    1.00  1.76e-09   6.7e-09  \n",
       "1                    0.99  2.48e-13  2.63e-12  \n",
       "2                    0.95  1.96e-11  4.67e-10  \n",
       "3                    0.93  1.96e-13  3.53e-11  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.set_option('display.max_colwidth', None)\n",
    "table_latex = eQTL_dep_table.to_latex(index = False,  escape=False,\n",
    "              column_format='p{0.7cm}|p{1.8cm}|p{1.7cm}|p{1.2cm}|p{1cm}|p{1cm}|p{1cm}|p{1cm}|p{1.3cm}|p{1.3cm}')\n",
    "table_latex = table_latex.replace(\"\\\\\\n\", \"\\\\ \\hline\\n\") # add lines between rows\n",
    "with open('/data/srlab/lrumker/MCSC_Project/cna-qtl/tables/supptable.trans_masked.tex','w') as tf:\n",
    "    tf.write(table_latex)\n",
    "eQTL_dep_table"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
